File: sdks/typescript/packages/policy/src/engine.ts
import type {
  PaymentPolicy,
  PolicyContext,
  PolicyDecision,
  SessionStats,
} from "./types";

interface PaymentRecord {
  amount: bigint;
  timestamp: number;
}

/**
 * Payment policy engine that evaluates payment requirements against
 * a configured policy and tracks session statistics.
 */
export class PaymentPolicyEngine {
  private readonly policy: PaymentPolicy;
  private readonly payments: PaymentRecord[] = [];
  private sessionStartTime: number;
  private readonly nowFn: () => number;

  /**
   * Create a new policy engine.
   * @param policy - The payment policy to enforce
   * @param options - Optional configuration (e.g. custom clock for testing)
   */
  constructor(
    policy: PaymentPolicy,
    options?: { now?: () => number },
  ) {
    this.policy = policy;
    this.nowFn = options?.now ?? (() => Date.now());
    this.sessionStartTime = this.nowFn();
  }

  /**
   * Evaluate a payment against the configured policy.
   * Returns { allowed: true } if the payment passes all checks,
   * or { allowed: false, reason } with the first failing rule.
   */
  async evaluate(requirements: {
    scheme: string;
    network: string;
    asset: string;
    amount: string;
    payTo: string;
  }): Promise<PolicyDecision> {
    const amount = BigInt(requirements.amount);
    const now = this.nowFn();
    const stats = this.computeStats(now);

    // Built-in rules (checked in order)
    const builtInResult = this.evaluateBuiltInRules(requirements, amount, stats, now);
    if (!builtInResult.allowed) {
      return builtInResult;
    }

    // Custom rules
    if (this.policy.customRules) {
      const context: PolicyContext = { requirements, session: stats };
      for (const rule of this.policy.customRules) {
        const decision = await rule.validate(context);
        if (!decision.allowed) {
          return { allowed: false, reason: `Custom rule "${rule.name}": ${decision.reason}` };
        }
      }
    }

    return { allowed: true };
  }

  /**
   * Record a successful payment. Call this after a payment is confirmed
   * so that session stats are updated for future evaluations.
   */
  recordPayment(amount: string): void {
    this.payments.push({
      amount: BigInt(amount),
      timestamp: this.nowFn(),
    });
  }

  /**
   * Get current session statistics.
   */
  getStats(): SessionStats {
    return this.computeStats(this.nowFn());
  }

  /**
   * Reset the engine, clearing all session history.
   */
  reset(): void {
    this.payments.length = 0;
    this.sessionStartTime = this.nowFn();
  }

  private evaluateBuiltInRules(
    requirements: { scheme: string; network: string; asset: string; amount: string; payTo: string },
    amount: bigint,
    stats: SessionStats,
    _now: number,
  ): PolicyDecision {
    // maxAmountPerPayment
    if (this.policy.maxAmountPerPayment !== undefined) {
      const max = BigInt(this.policy.maxAmountPerPayment);
      if (amount > max) {
        return {
          allowed: false,
          reason: `Amount ${amount} exceeds max per payment ${max}`,
        };
      }
    }

    // maxAmountPerSession
    if (this.policy.maxAmountPerSession !== undefined) {
      const max = BigInt(this.policy.maxAmountPerSession);
      if (stats.totalAmountPaid + amount > max) {
        return {
          allowed: false,
          reason: `Cumulative session amount would be ${stats.totalAmountPaid + amount}, exceeding max ${max}`,
        };
      }
    }

    // maxAmountPerDay
    if (this.policy.maxAmountPerDay !== undefined) {
      const max = BigInt(this.policy.maxAmountPerDay);
      if (stats.amountPaidToday + amount > max) {
        return {
          allowed: false,
          reason: `Daily amount would be ${stats.amountPaidToday + amount}, exceeding max ${max}`,
        };
      }
    }

    // maxPaymentsPerHour
    if (this.policy.maxPaymentsPerHour !== undefined) {
      if (stats.paymentsThisHour >= this.policy.maxPaymentsPerHour) {
        return {
          allowed: false,
          reason: `Payments this hour (${stats.paymentsThisHour}) would exceed max ${this.policy.maxPaymentsPerHour}`,
        };
      }
    }

    // allowedRecipients
    if (this.policy.allowedRecipients !== undefined && this.policy.allowedRecipients.length > 0) {
      const normalized = requirements.payTo.toLowerCase();
      const allowed = this.policy.allowedRecipients.map((r) => r.toLowerCase());
      if (!allowed.includes(normalized)) {
        return {
          allowed: false,
          reason: `Recipient ${requirements.payTo} is not in the allowed list`,
        };
      }
    }

    // blockedRecipients
    if (this.policy.blockedRecipients !== undefined && this.policy.blockedRecipients.length > 0) {
      const normalized = requirements.payTo.toLowerCase();
      const blocked = this.policy.blockedRecipients.map((r) => r.toLowerCase());
      if (blocked.includes(normalized)) {
        return {
          allowed: false,
          reason: `Recipient ${requirements.payTo} is blocked`,
        };
      }
    }

    // allowedNetworks
    if (this.policy.allowedNetworks !== undefined && this.policy.allowedNetworks.length > 0) {
      if (!this.policy.allowedNetworks.includes(requirements.network)) {
        return {
          allowed: false,
          reason: `Network ${requirements.network} is not in the allowed list`,
        };
      }
    }

    // allowedSchemes
    if (this.policy.allowedSchemes !== undefined && this.policy.allowedSchemes.length > 0) {
      if (!this.policy.allowedSchemes.includes(requirements.scheme)) {
        return {
          allowed: false,
          reason: `Scheme ${requirements.scheme} is not in the allowed list`,
        };
      }
    }

    // allowedAssets
    if (this.policy.allowedAssets !== undefined && this.policy.allowedAssets.length > 0) {
      const normalized = requirements.asset.toLowerCase();
      const allowed = this.policy.allowedAssets.map((a) => a.toLowerCase());
      if (!allowed.includes(normalized)) {
        return {
          allowed: false,
          reason: `Asset ${requirements.asset} is not in the allowed list`,
        };
      }
    }

    return { allowed: true };
  }

  private computeStats(now: number): SessionStats {
    const oneHourAgo = now - 60 * 60 * 1000;
    const oneDayAgo = now - 24 * 60 * 60 * 1000;

    let totalAmountPaid = 0n;
    let amountPaidToday = 0n;
    let paymentsThisHour = 0;

    for (const payment of this.payments) {
      totalAmountPaid += payment.amount;

      if (payment.timestamp > oneDayAgo) {
        amountPaidToday += payment.amount;
      }

      if (payment.timestamp > oneHourAgo) {
        paymentsThisHour++;
      }
    }

    return {
      totalAmountPaid,
      paymentCount: this.payments.length,
      paymentsThisHour,
      amountPaidToday,
      startTime: this.sessionStartTime,
    };
  }
}

File: sdks/go/policy/engine.go
package policy

import (
	"fmt"
	"math/big"
	"strings"
	"sync"
	"time"
)

// Engine evaluates payment requests against a policy and tracks session state.
type Engine struct {
	mu       sync.Mutex
	policy   *PaymentPolicy
	stats    SessionStats
	payments []paymentRecord
}

type paymentRecord struct {
	amount *big.Int
	time   time.Time
}

// NewEngine creates a new policy engine with the given policy.
func NewEngine(policy *PaymentPolicy) *Engine {
	return &Engine{
		policy: policy,
		stats: SessionStats{
			TotalAmountPaid: new(big.Int),
			AmountPaidToday: new(big.Int),
			StartTime:       time.Now(),
		},
	}
}

// Evaluate checks whether a payment is permitted by the policy.
func (e *Engine) Evaluate(scheme, network, asset, amount, payTo string) Decision {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.pruneHourlyPayments()

	ctx := &Context{
		Scheme:  scheme,
		Network: network,
		Asset:   asset,
		Amount:  amount,
		PayTo:   payTo,
		Session: e.currentStats(),
	}

	// Check per-payment amount limit.
	if e.policy.MaxAmountPerPayment != "" {
		if d := checkAmountLimit(amount, e.policy.MaxAmountPerPayment, "per-payment limit"); !d.Allowed {
			return d
		}
	}

	// Check per-session cumulative limit.
	if e.policy.MaxAmountPerSession != "" {
		projected := new(big.Int).Set(e.stats.TotalAmountPaid)
		paymentAmt, ok := new(big.Int).SetString(amount, 10)
		if !ok {
			return Deny(fmt.Sprintf("invalid amount: %s", amount))
		}
		projected.Add(projected, paymentAmt)
		limit, ok := new(big.Int).SetString(e.policy.MaxAmountPerSession, 10)
		if !ok {
			return Deny(fmt.Sprintf("invalid session limit: %s", e.policy.MaxAmountPerSession))
		}
		if projected.Cmp(limit) > 0 {
			return Deny(fmt.Sprintf("would exceed session limit: %s + %s > %s",
				e.stats.TotalAmountPaid.String(), amount, e.policy.MaxAmountPerSession))
		}
	}

	// Check per-day cumulative limit.
	if e.policy.MaxAmountPerDay != "" {
		projected := new(big.Int).Set(e.stats.AmountPaidToday)
		paymentAmt, ok := new(big.Int).SetString(amount, 10)
		if !ok {
			return Deny(fmt.Sprintf("invalid amount: %s", amount))
		}
		projected.Add(projected, paymentAmt)
		limit, ok := new(big.Int).SetString(e.policy.MaxAmountPerDay, 10)
		if !ok {
			return Deny(fmt.Sprintf("invalid daily limit: %s", e.policy.MaxAmountPerDay))
		}
		if projected.Cmp(limit) > 0 {
			return Deny(fmt.Sprintf("would exceed daily limit: %s + %s > %s",
				e.stats.AmountPaidToday.String(), amount, e.policy.MaxAmountPerDay))
		}
	}

	// Check hourly payment count.
	if e.policy.MaxPaymentsPerHour > 0 {
		if e.stats.PaymentsThisHour >= e.policy.MaxPaymentsPerHour {
			return Deny(fmt.Sprintf("would exceed hourly payment limit: %d >= %d",
				e.stats.PaymentsThisHour, e.policy.MaxPaymentsPerHour))
		}
	}

	// Check allowed recipients.
	if len(e.policy.AllowedRecipients) > 0 {
		if !containsCaseInsensitive(e.policy.AllowedRecipients, payTo) {
			return Deny(fmt.Sprintf("recipient not in allowlist: %s", payTo))
		}
	}

	// Check blocked recipients.
	if len(e.policy.BlockedRecipients) > 0 {
		if containsCaseInsensitive(e.policy.BlockedRecipients, payTo) {
			return Deny(fmt.Sprintf("recipient is blocked: %s", payTo))
		}
	}

	// Check allowed networks.
	if len(e.policy.AllowedNetworks) > 0 {
		if !containsCaseInsensitive(e.policy.AllowedNetworks, network) {
			return Deny(fmt.Sprintf("network not allowed: %s", network))
		}
	}

	// Check allowed schemes.
	if len(e.policy.AllowedSchemes) > 0 {
		if !containsCaseInsensitive(e.policy.AllowedSchemes, scheme) {
			return Deny(fmt.Sprintf("scheme not allowed: %s", scheme))
		}
	}

	// Check allowed assets.
	if len(e.policy.AllowedAssets) > 0 {
		if !containsCaseInsensitive(e.policy.AllowedAssets, asset) {
			return Deny(fmt.Sprintf("asset not allowed: %s", asset))
		}
	}

	// Check custom rules.
	for _, rule := range e.policy.CustomRules {
		if d := rule.Validate(ctx); !d.Allowed {
			return Deny(fmt.Sprintf("custom rule %q: %s", rule.Name, d.Reason))
		}
	}

	return Allow()
}

// RecordPayment records a successful payment to update session statistics.
// The amount must be in the smallest unit (e.g., wei, lamports).
func (e *Engine) RecordPayment(amount string) {
	e.mu.Lock()
	defer e.mu.Unlock()

	amt, ok := new(big.Int).SetString(amount, 10)
	if !ok {
		return
	}

	now := time.Now()
	e.stats.TotalAmountPaid.Add(e.stats.TotalAmountPaid, amt)
	e.stats.AmountPaidToday.Add(e.stats.AmountPaidToday, amt)
	e.stats.PaymentCount++
	e.payments = append(e.payments, paymentRecord{amount: amt, time: now})

	e.pruneHourlyPayments()
}

// Reset clears all session statistics.
func (e *Engine) Reset() {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.stats = SessionStats{
		TotalAmountPaid: new(big.Int),
		AmountPaidToday: new(big.Int),
		StartTime:       time.Now(),
	}
	e.payments = nil
}

// Stats returns a snapshot of current session statistics.
func (e *Engine) Stats() SessionStats {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.pruneHourlyPayments()
	return *e.currentStats()
}

// currentStats returns a copy of the current stats (must be called with lock held).
func (e *Engine) currentStats() *SessionStats {
	return &SessionStats{
		TotalAmountPaid:  new(big.Int).Set(e.stats.TotalAmountPaid),
		PaymentCount:     e.stats.PaymentCount,
		PaymentsThisHour: e.stats.PaymentsThisHour,
		AmountPaidToday:  new(big.Int).Set(e.stats.AmountPaidToday),
		StartTime:        e.stats.StartTime,
	}
}

// pruneHourlyPayments removes payments older than one hour and updates the count.
func (e *Engine) pruneHourlyPayments() {
	cutoff := time.Now().Add(-1 * time.Hour)
	count := 0
	for _, p := range e.payments {
		if p.time.After(cutoff) {
			count++
		}
	}
	e.stats.PaymentsThisHour = count
}

// checkAmountLimit compares amount against a limit, both as big.Int strings.
func checkAmountLimit(amount, limit, label string) Decision {
	amt, ok := new(big.Int).SetString(amount, 10)
	if !ok {
		return Deny(fmt.Sprintf("invalid amount: %s", amount))
	}
	lim, ok := new(big.Int).SetString(limit, 10)
	if !ok {
		return Deny(fmt.Sprintf("invalid %s: %s", label, limit))
	}
	if amt.Cmp(lim) > 0 {
		return Deny(fmt.Sprintf("amount %s exceeds %s of %s", amount, label, limit))
	}
	return Allow()
}

// containsCaseInsensitive checks if the slice contains the target (case-insensitive).
func containsCaseInsensitive(list []string, target string) bool {
	lower := strings.ToLower(target)
	for _, item := range list {
		if strings.ToLower(item) == lower {
			return true
		}
	}
	return false
}
